### PR TITLE
Add support for attributes and events

### PR DESCRIPTION
**Pull request by The-inside-man**
## Summary

- Add support for Attributes and Events
- Add test cases to address new changes

## Test plan

Compatibility tests

## Issues
- OASIS-7798

### CODE REVIEW

**Review by keppel2 — APPROVED** [review]

File: core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfigService.java [hunk 1]
```diff
@@ -215,4 +217,40 @@ String getExperimentFeatureKey(String experimentId) {
 
         return featureVariableKeyMap;
     }
+
+    @VisibleForTesting
+    Map<String, OptimizelyAttribute> getAttributesMap(List<Attribute> attributes) {
```
Comment (jaeopt) [inline core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfigService.java:222]: We do not use this attributesMap and eventsMap in OptimizelyConfig. I think we can delete these.

File: core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfig.java [hunk 2]
```diff
@@ -53,6 +57,32 @@ public OptimizelyConfig(Map<String, OptimizelyExperiment> experimentsMap,
...
+        this(experimentsMap, featuresMap, revision, sdkKey, environmentKey, attributes, events, null);
+    }
+
+    public OptimizelyConfig(Map<String, OptimizelyExperiment> experimentsMap,
+                            Map<String, OptimizelyFeature> featuresMap,
+                            String revision,
+                            String sdkKey,
+                            String environmentKey,
+                            List<Attribute> attributes,
+                            List<EventType> events,
```
Comment (jaeopt) [inline core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfig.java:74]:
```suggestion
                            List<OptimizelyAttribute> attributes,
                            List<OptimizelyEvent> events,
```

File: core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfig.java [hunk 3]
```diff
@@ -61,6 +91,10 @@ public OptimizelyConfig(Map<String, OptimizelyExperiment> experimentsMap,
         return featuresMap;
     }
 
+    public List<Attribute> getAttributes() { return attributes; }
```
Comment (jaeopt) [inline core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfig.java:94]:
```suggestion
    public List<OptimizelyAttribute> getAttributes() { return attributes; }
```

File: core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfig.java [hunk 4]
```diff
@@ -61,6 +91,10 @@ public OptimizelyConfig(Map<String, OptimizelyExperiment> experimentsMap,
...
+
+    public List<EventType> getEvents() { return events; }
```
Comment (jaeopt) [inline core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfig.java:96]:
```suggestion
    public List<OptimizelyEvent> getEvents() { return events; }
```

**Review by jaeopt — CHANGES_REQUESTED** [review]
OptimizelyAttribute and OptimizelyEvent not passed

File: core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfigService.java (hunk 1 above)
Comment (The-inside-man) [reply core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfigService.java]: removed

File: core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfig.java (hunk 2 above)
Comment (The-inside-man) [reply core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfig.java]: Changed

File: core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfig.java (hunk 3 above)
Comment (The-inside-man) [reply core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfig.java]: Changed

File: core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfig.java (hunk 4 above)
Comment (The-inside-man) [reply core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfig.java]: Changed

**Review by jaeopt — APPROVED** [review]
LGTM

**Comment by keppel2** [comment]
Changes are fine.

**Comment by The-inside-man** [comment]
@msohailhussain - Are changes needed in the compatibility suite for this to pass as well? Looks like the tests there are failing due to attributes and events being present.

**Comment by The-inside-man** [comment]
> @msohailhussain - Are changes needed in the compatibility suite for this to pass as well? Looks like the tests there are failing due to attributes and events being present.

It looks like a small change need to be made to get_optimizely_config.feature in the compatibility suite to expect attributes and events.

**Comment by keppel2** [comment]
> > @msohailhussain - Are changes needed in the compatibility suite for this to pass as well? Looks like the tests there are failing due to attributes and events being present.
> 
> It looks like a small change need to be made to get_optimizely_config.feature in the compatibility suite to expect attributes and events.

Then it won't pass for master.

**Comment by keppel2** [comment]
> > > @msohailhussain - Are changes needed in the compatibility suite for this to pass as well? Looks like the tests there are failing due to attributes and events being present.
> > 
> > 
> > It looks like a small change need to be made to get_optimizely_config.feature in the compatibility suite to expect attributes and events.
> 
> Then it won't pass for master.

Just coordinate the two. I did something similar in Go (a refactor).

**Comment by The-inside-man** [comment]
Passing all tests except ones looking for the Audiences implementation. Will be added in separate PR.
link to FSC for more details: 

https://travis-ci.com/github/optimizely/fullstack-sdk-compatibility-suite/builds/232501967

**Comment by The-inside-man** [comment]
Closing out as Audiences PR has all updated code for Attributes, Events and Audiences.

**Pull request closed without merging**